src/auth/permissions/index.ts

```diff
@@ -0,0 +1,173 @@
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
+        const permissionId = permissionIds[0].permission_id;
+
+        const insertPermission = await client('access_role_permission')
+            .returning('*')
+            .insert(
+              client.select({
+                ['access_role_id']: roleId,
+                ['permission_id']: permissionId,
+              })
+                .whereNotExists(client('access_role_permission').where({
+                  ['access_role_id']: roleId,
+                  ['permission_id']: permissionId,
+                }))
+            );
+        if (insertPermission.length === 0) {
+          throw new Error('Permission entry is already associated to this role');
+        }
+        client.destroy();
+        return insertPermission;
+      } else {
+        client.destroy();
+        throw new Error('Permission entry does not exists, please use grantNew method');
+      }
+    },
+    grantNew: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 0) {
+        const createPermissionId = await client('permission')
+            .returning('permission_id')
+            .insert({
+              ['permission_entity']: resource,
+              ['permission_level']: permissionLevel,
+              ['access_type']: access,
+            });
+
+        const permissionId = createPermissionId[0];
+        const insertPermission = await client('access_role_permission')
+            .returning('*')
+            .insert({
+              ['access_role_id']: roleId,
+              ['permission_id']: permissionId,
+            });
+        client.destroy();
+        return insertPermission;
+      } else {
+        throw new Error('Permission already exists, please use grantExisting method');
+        client.destroy();
+      }
+    },
+    revoke: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
+        const permissionId = permissionIds[0].permission_id;
+
+        const deleteRow = await client('access_role_permission')
+            .where({
+              ['access_role_id']: roleId,
+              ['permission_id']: permissionId,
+            })
+            .del();
+
+        if (deleteRow === 1) return deleteRow;
+        throw new Error('Permission entry is not linked to role');
+      }
```

Comment:
I can't see whether the brackets line up here, but if `permissionIds.length !== 1`, it looks like we just do nothing and return `undefined`. Is that right?